### DISCUSSION TITLE
test for user logged in server cache issue

### DISCUSSION THREAD

**Discussion by rogerg1967** [openstore-ecommerce/OpenStore#254] [Q&A]
I am using the following to check if a user has a specific role and if a product has attached documents to display the downloads.
@if (DotNetNuke.Security.PortalSecurity.IsInRole ("abc") == true)  {
            if (@product.Docs.Count >= 1)
The issue I am having is that when a user logs in or out I have to clear the server cache before the razor view updates. Can you advise if there is a way to resolve this or a better way to achieve what I want?
Thanks
Roger

**Comment by DNNMonster** [comment]
It's likely a cache key issue.  You need to set a different cache key for any unique razor file rendering.  If your display targets a specific role then you would want to ensure your cache key includes a role id or role name to differentiate it from the unauthenticated users or users in different roles.  You'll just have to modify your custom theme templates so the params being used in the calls to the render functions reflect above.

**Reply by rogerg1967** [reply]
How do we specify the cache key in a template?

**Comment by DNNMonster** [comment]
There may be other approaches that @leedavi can share but one can solve this with a pre-processor meta token.  We added a pre-processor meta token to ProductDisplayAjax page and changed the moduleId param by appending a value(userid or roleid).  I don't think it involved other code changes but I tackled this a while ago so you'll need to test.  https://github.com/nbrightproject/NBrightTS/blob/c05d6d4a4a1fe31d6e70d0dbf3e7603b6db42a49/render/RazorEngineTokens.cs#L62

**Comment by Rodien** [comment]
I'm running into something similar and running in debug mode seems to help. 

I will try the pre-process token

**Reply by rogerg1967** [reply]
@Rodien it would be good to know if it resolves your issue. I will report back what I find as well.
Regards

**Comment by rogerg1967** [comment]
Can anyone advise how to add the pre-process function in a template as I am not sure how to do this?

**Comment by leedavi** [comment]
I'm unsure the pre-process will work.  It's usually is for SQL selections, so the selection itself can avoid caching.

The problem is that the user is not recognized as a cache trigger, for obvious reasons that in theory users can be unlimited, so caching on them would not help high load websites. 
In the past we have made ajax calls from the client to display user specific data, but there is a price to that action.

I don't like running in debug mode, but it is a way to deal with it.  If the website doesn't have a large amount of traffic, it is a workable  solution.

So, I think you have 2 options that can work.

1.  Leave it in debug mode.  There are potential memory leak issues with debug mode, but it has been much improved from the early days and is a low risk, easy solution.
2.  Create a plugin to do an ajax call for the specific functionality required.  This is the best solution, but it does take time and money.  It also means each page load will have a call.  To offset this you can have a main public page and then a "members" page to display special details. 

Option 1 should be OK with a low volume website.  If you find it slow then option 2 is the option.

**Comment by leedavi** [comment]
Ideally, it would be nice to have a Pre-Process token that specified any additional cache values or turned on/off cache for that template.
